Switched the project to C++

kotachang opened the pull request:
### Description
Switched the project language to C++. Functionality should remain the same to the user (phone or python script)

### Changes in the codebase
CMakeLists.txt modified to accomodate changes to the files
soundinator_main.c -> soundinator_main.cpp
nimble/ble functions inside soundinator_main.c -> restructured as the Nimbler class in nimbler.cpp and nimbler.hpp
code sequence (flow of function calls) remains the same

### Testing instructions
1. Build and flash from the original branch
2. Run the python_client.py script as the following: `python python_client.py --name Soundinator face` (this specifies the device name as Soundinator, and face is the hex representation of the characteristic uuid we want to get notified from)
3. Take note of the output from the script
4. Build and flash from this branch (with the C++ code)
5. Repeat steps 2-3 and check that the output is identical to the main branch

kotachang committed:
Starting to convert main.c to cpp

Created soundinator_main.cpp file for transition. File is empty and no other changes have been made.

kotachang committed:
Started cpp and hpp structure, modified CMakeLists accordingly

Removed main .c code, added the beginning of the main code to .cpp, created Nimbler class, and modified CMakeLists to match that. Build and flash tested but BLE functionalities not yet tested

kotachang committed:
Nimbler class init() function implemented

First integration of the nimble initialization sequence as init() function in Nimbler class. struct definitions and callbacks for defining the gap & gatt service had to be restructured to avoid assigning addresses of rvalues. Build tested but BLE connection functionality not tested yet.

kotachang committed:
Nimbler ble advertisement implemented

Implemented upto the ble advertisement but does not actually call the nimbler advertisement function yet. All callbacks have now been moved to private after syntax has been figured out. Build and flash tested but BLE functionality not tested yet.

kotachang committed:
Client subscription and notification handling added and tested

Now at the same spot as the original c code. The client Python script can subscribe, get notified, and write to the ESP32. Needed to change all the gatt characteristic struct definition to static to allow the nimble stack to access it properly.

dvirhilu commented on ESP32/main/src/soundinator_main.c:
goodbye old friend

dvirhilu commented on ESP32/main/src/soundinator_main.cpp:
```diff
@@ -0,0 +1,38 @@
+#include "nvs_flash.h"
+#include "esp_log.h"
+#include "esp_nimble_hci.h"
+#include "nimble/nimble_port.h"
```
Are all of the nimble includes required for the main fail? If possible would be better to only include them inside nimbler.hpp for better encapsulation.

Maybe to clarify a little bit, I just meant that we should only "#include" libraries if the code in the file uses them (if they are used by downstream code, they should be included in the downstream file).

dvirhilu commented on ESP32/main/src/soundinator_main.cpp:
```diff
@@ -0,0 +1,38 @@
+#include "sdkconfig.h"
+#include "nimbler.hpp"
+
+// tag for logs
```
fix comment to follow standard

dvirhilu commented on ESP32/main/src/soundinator_main.cpp:
```diff
@@ -0,0 +1,38 @@
+const char* tag = "Main";
+
+/*
+* Follow programming sequence from https://docs.espressif.com/projects/esp-idf/en/v5.3.1/esp32/api-reference/bluetooth/nimble/index.html
```
update this comment - since changing the format to c++ it now diverges from the example. The citation link is probably better in Nimbler.cpp and then update the comment to just describe what this function is doing.

dvirhilu commented on ESP32/main/inc/nimbler.hpp:
```diff
@@ -0,0 +1,56 @@
+
+#define N_SERVICES              (2)
+#define N_CHARACTERISTICS       (3)
+#define service_uuid_val        (0xBEEF)
```
allergic reaction

dvirhilu commented on ESP32/main/inc/nimbler.hpp:
```diff
@@ -0,0 +1,56 @@
+
+class Nimbler {
+    public:
+        const char* device_name;
```
might be better off as a private member with a getter function to guarantee immutability. Need to look into exact const char * semantics to check whether this is immutable as is.

dvirhilu commented on ESP32/main/inc/nimbler.hpp:
```diff
@@ -0,0 +1,56 @@
+class Nimbler {
+    public:
+        const char* device_name;
+        static constexpr char* tag = "Nimbler"; // tag used for logs
```
Again, might be better as a private member or as a macro

dvirhilu replied:
also look into why constexpr vs const

dvirhilu commented on ESP32/main/src/nimbler.cpp:
```diff
@@ -0,0 +1,203 @@
+    },
+};
+
+/* gatt service structure definition */
```
capitalization

dvirhilu commented on ESP32/main/src/nimbler.cpp:
```diff
@@ -0,0 +1,203 @@
+    .value = device_write_val
+};
+
+/* gatt characteristic structure definition */
```
capitalization

dvirhilu commented on ESP32/main/src/nimbler.cpp:
```diff
@@ -0,0 +1,203 @@
+uint16_t Nimbler::conn_hndl = 0;
+uint16_t Nimbler::attr_hndl_audio = 0;
+
+/* uuids for main ble service and characteristics */
```
capitalization

dvirhilu commented on ESP32/main/src/nimbler.cpp:
```diff
@@ -0,0 +1,203 @@
+/* Constructor */
+Nimbler::Nimbler(const char* device_name) {
+    this->device_name = device_name;
+    ble_addr_type = 0; // Determines the best address type automatically
```
uncapitalize inline comment

dvirhilu commented on ESP32/main/src/nimbler.cpp:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    /* Initialize the NimBLE host configuration. */
+    ble_hs_cfg.sync_cb = _ble_app_on_sync;   // Initialize advertisement only after the host & controller are synced
```
inline comment uncapitalize

dvirhilu commented on ESP32/main/src/nimbler.cpp:
```diff
@@ -0,0 +1,203 @@
+void Nimbler::_ble_task(void *param)
+{
+    ESP_LOGI(tag, "BLE Host Task Started");
+    nimble_port_run(); // This function will return only when nimble_port_stop() is executed
```
uncapitalize inline comment

dvirhilu commented on ESP32/main/src/nimbler.cpp:
```diff
@@ -0,0 +1,203 @@
+/* Start of BLE application after NimBLE host & controller are synced */
+void Nimbler::_ble_app_on_sync(void)
+{
+    ble_hs_id_infer_auto(0, &ble_addr_type); // Determines the best address type automatically
```
inline comment uncapitalize

dvirhilu commented on ESP32/main/src/nimbler.cpp:
```diff
@@ -0,0 +1,203 @@
+    struct ble_hs_adv_fields fields;
+    const char *device_name;
+    memset(&fields, 0, sizeof(fields));
+    device_name = ble_svc_gap_device_name(); // Read the BLE device name
```
uncapitalize

kotachang committed:
Modified based on PR review

Modified comments for better standardization, moved device_name to a private member.

kotachang committed:
PR comment address - Moved log tag to private

dvirhilu reviewed (approved):
looks goo!

jackysioo commented on ESP32/main/inc/nimbler.hpp:
```diff
@@ -0,0 +1,56 @@
+#define N_CHARACTERISTICS       (3)
+#define service_uuid_val        (0xBEEF)
+#define device_read_uuid_val    (0xBEBE)
+#define device_write_val        (0xFACE)
```
rename `device_write_val` to `device_write_uuid_val` for consistency

kotachang committed:
PR comment address - Renamed device_write_val to devcie_write_uuid_val

jackysioo commented on ESP32/main/inc/nimbler.hpp:
```diff
@@ -0,0 +1,56 @@
+        static void _ble_app_on_reset(int reason);
+        static void _ble_app_advertise();
+        static int _ble_gap_event(struct ble_gap_event *event, void *arg);
+        static int _device_write(uint16_t conn_handle, uint16_t attr_handle, struct ble_gatt_access_ctxt *ctxt, void *arg);
+        static int _device_read(uint16_t con_handle, uint16_t attr_handle, struct ble_gatt_access_ctxt *ctxt, void *arg);
```
rename `con_handle` to `conn_handle` in `_device_read`

kotachang replied:
conn_handle is specific to the BLE connection and not a connection to a specific attribute (ex. device_read)

kotachang committed:
PR comment address - fix conn_handle typo

jackysioo commented:
Looks good!
![soundinator_c](https://github.com/user-attachments/assets/f3e9fa20-865f-41b0-bfac-c2d3502e57b8)
![soundinator_cpp](https://github.com/user-attachments/assets/a3377355-a89d-4977-964c-5ac19032d8b3)
